Use XCTestRun to execute tests and report results

modocache opened the pull request:
## What's in this pull request?

A major goal for swift-corelibs-xctest is API parity with Apple XCTest. This adds the largest missing API in swift-corelibs-xctest: `XCTestRun`.

In Apple XCTest, `XCTestRun` is responsible for keeping track of the result of a test run. It's an integral part of how Apple XCTest works. swift-corelibs-xctest, on the other hand, used a global array of `XCTRun` structs to keep track of how many tests passed/failed.

While it may have been possible to tack on `XCTestRun` to the swift-corelibs-xctest mechanism for failure reporting, this commit instead fully integrates it. As a result, the changes are widespread: gone is `XCTFailureHandler`, `XCTRun`, and other internal structures. In their place, welcome the Apple XCTest public APIs: the `XCTest` abstract class, `XCTestRun`, and its subclasses `XCTestCaseRun` and `XCTestSuiteRun`.

In conjunction with the new `XCTestSuite`-related observation methods from #84, test reporting is now done exclusively through `XCTestObservation`. As a result, test output is now nearly identical to Apple XCTest.

## Why merge this pull request?

- Our goal for Swift 3 is API parity, and this pull request implements a ton of the Apple XCTest API.
- This pull request implements test output that is nearly identical to Apple XCTest, including announcements about test suites starting and stopping.

## What downsides are there to merging this pull request?

- This is a massive overhaul of swift-corelibs-xctest internals. It's hard to review such a large set of changes. I am sorry about this, and merging #84 first should slim this pull request down a bit--but even then, this is huge. In my defense, `XCTestRun` is a major part of the Apple XCTest API; introducing it to swift-corelibs-xctest requires a lot of changes.
- I recommend looking at the changes to `Tests/Functional` first. Note that these changes report test suites, as well as start and stop times for each test suite/case.
- A lot of the existing Swift code is much more "idiomatic" than the code here, which introduces two "abstract" base classes: `XCTest` and `XCTestRun`. Like it or not, this is the Apple XCTest API, which has been built for Objective-C since day one. We'll leave it to the third-party testing frameworks to build APIs more suitable to Swift.

modocache commented:
@mike-ferris-apple @briancroom Would love to hear your feedback on this one! :wave:

mike-ferris-apple commented:
I looked at the prior one to add suites to the observation protocol. It may be a few days before I can look at this one in any detail.

mike-ferris-apple commented on Sources/XCTest/XCAbstractTest.swift:
```diff
@@ -33,4 +64,8 @@ public class XCTest {
+
+    // FIXME: This initializer is required due to a Swift compiler bug on Linux.
+    //        It should be removed once the bug is fixed.
+    public init() {}
```
Should reference the specific bug here...

modocache replied:
Oh, good point. I have a feeling this is subtly different from SR-1129, but I'll double check.

gruberjuliannama replied:
// Not all expectations were fulfilled. Append a failure
              // to the array of expectation-based failures.
              let descriptions = unfulfilledDescriptions.joined(separator: ", ")
 -            let failure = XCTFailure(
 -                message: "Exceeded timeout of \(timeout) seconds, with unfulfilled expectations: \(descriptions)",
 -                failureDescription: "Asynchronous wait failed",
 -                expected: true,
 -                file: file,
 -                line: line)
 -            if let failureHandler = XCTFailureHandler {
 -                failureHandler(failure)
 -            }
 +            recordFailureWithDescription(
 +                "Asynchronous wait failed - Exceeded timeout of \(timeout) seconds, with unfulfilled expectations: \(descriptions)",
 +                inFile: String(file),

mike-ferris-apple commented:
I've looked through this. This actually seems to clean up a lot of stuff in addition to getting more API parity. I like that the printing is now handled by an observer, and, in general things seem to be better encapsulated and separated.

I don't have the bandwidth to truly assess for functional equivalence with Xcode's XCTest behavior, but the API all seems good and I saw no obvious issues in the implementation.

modocache commented:
@mike-ferris-apple Excellent! Could you try kicking off the tests?

@briancroom said he might have time to review this soon so I'll wait for that before considering merging.

ddunbar commented:
@swift-ci please test

gruberjuliannama commented:
// Not all expectations were fulfilled. Append a failure
// to the array of expectation-based failures.
let descriptions = unfulfilledDescriptions.joined(separator: ", ")
- let failure = XCTFailure(
- message: "Exceeded timeout of \(timeout) seconds, with unfulfilled
expectations: \(descriptions)",
- failureDescription: "Asynchronous wait failed",
- expected: true,
- file: file,
- line: line)
- if let failureHandler = XCTFailureHandler {
- failureHandler(failure)
- }
+ recordFailureWithDescription(
+ "Asynchronous wait failed - Exceeded timeout of \(timeout) seconds, with
unfulfilled expectations: \(descriptions)",
+ inFile: String(file),

2016-04-03 18:32 GMT+02:00 Brian Gesiak <notifications@github.com>:

> @mike-ferris-apple <https://github.com/mike-ferris-apple> Excellent!
> Could you try kicking off the tests?
>
> @briancroom <https://github.com/briancroom> said he might have time to
> review this soon so I'll wait for that before considering merging.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86#issuecomment-205007319>
>

briancroom commented on Sources/XCTest/XCAbstractTest.swift:
```diff
@@ -26,6 +26,37 @@ public class XCTest {
 
+    /// The `XCTestRun` subclass that will be instantiated when the test is run
+    /// to hold the test's results. Must be overridden by subclasses.
+    public var testRunClass: AnyClass {
```
Starting in the version of `XCTest.framework` that ships with Xcode 7.3, this property has a type of `AnyClass?`. If we want to strictly follow Apple XCTest here, we should use that. Alternatively, I'm wondering if it would be reasonable to adjust this to have a type of `XCTestRun.Type` (or `XCTestRun.Type?`) which would be considerably Swiftier while (IMO) not breaking the intention of the original API.

modocache replied:
Nice catch. Updated to use `AnyClass?` here. Submitted https://bugs.swift.org/browse/SR-1164 (with radar) to track changing the type in the SDK overlay.

briancroom commented on Sources/XCTest/XCAbstractTest.swift:
```diff
@@ -26,6 +26,37 @@ public class XCTest {
+
+    /// The test run object that executed the test, an instance of
+    /// testRunClass. If the test has not yet been run, this will be nil.
+    /// - Note: FIXME: This property is meant to be `internal(set)`. It is
```
Curiosity: why would this (ideally) be `internal(set)` as opposed to `private(set)`?

modocache replied:
Another good catch, thanks! Updated the documentation.

briancroom commented on Sources/XCTest/XCAbstractTest.swift:
```diff
@@ -26,6 +26,37 @@ public class XCTest {
+
+    /// Creates an instance of the `testRunClass` and passes it as a parameter
+    /// to `performTest()`.
+    public func runTest() {
```
Swift 3 API import change: `public func runTest()` becomes `public func run()`

modocache replied:
Are you sure you're not referring to `-[XCTest run]`, a deprecated method? https://github.com/modocache/XCTest-Public-Headers/blob/4b638808d0a742f03d4613f71e2eb03f3118498d/Developer/Platforms/iPhoneOS.platform/Developer/Library/Frameworks/XCTest.framework/Headers/Headers/XCAbstractTest.h#L87-L91

modocache replied:
Uh-oh, looks like we have a problem: https://github.com/modocache/XCTest.swift/blob/5b5a152150bacea9c373dcfb16a8384563cd4419/XCTest.swift#L18-L20

How does Swift determine which version to use here. I think we may need to file a Swift bug...

modocache replied:
https://bugs.swift.org/browse/SR-1165

briancroom replied:
Indeed - a quick test indicates that the compiler currently flags usages of these two methods as ambiguous :frowning: Thanks for filing the issue.

briancroom commented on Sources/XCTest/XCTAssert.swift:
```diff
@@ -87,8 +87,12 @@ private func _XCTEvaluateAssertion(assertion: _XCTAssertion, @autoclosure messag
     default:
-        if let handler = XCTFailureHandler {
-            handler(XCTFailure(message: message(), failureDescription: result.failureDescription(assertion), expected: result.expected, file: file, line: line))
+        if let currentTestCase = XCTCurrentTestCase {
```
:+1:

gruberjuliannama replied:
test_hasInitializedValue' passed \(\d+\.\d+ seconds\).
func test_hasInitializedValue() {
XCTAssertEqual(value, 1)
value += 1
}
-// CHECK: Test Case
'NewInstanceForEachTestTestCase.test_hasInitializedValueInAnotherTest'
started.
+// CHECK: Test Case
'NewInstanceForEachTestTestCase.test_hasInitializedValueInAnotherTest'
started at \d+:\d+:\d+\.\d+
// CHECK: Test Case
'NewInstanceForEachTestTestCase.test_hasInitializedValueInAnotherTest'
passed \(\d+\.\d+ seconds\).
func test_hasInitializedValueInAnotherTest() {
XCTAssertEqual(value, 1)
}
}
-// CHECK: Executed 2 tests, with 0 failures \(0 unexpected\) in \d+\.\d+
\(\d+\.\d+\) seconds
+// CHECK: Test Suite 'NewInstanceForEachTestTestCase' passed at
\d+:\d+:\d+\.\d+
+// CHECK: \t Executed 2 tests, with 0 failures \(0 unexpected\) in \d+\.\d+

2016-04-03 21:13 GMT+02:00 Brian Croom <notifications@github.com>:

> In Sources/XCTest/XCTAssert.swift
> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58315746>
> :
>
> > @@ -87,8 +87,12 @@ private func _XCTEvaluateAssertion(assertion: _XCTAssertion, @autoclosure messag
> >      case .Success:
> >          return
> >      default:
> > -        if let handler = XCTFailureHandler {
> > -            handler(XCTFailure(message: message(), failureDescription: result.failureDescription(assertion), expected: result.expected, file: file, line: line))
> > +        if let currentTestCase = XCTCurrentTestCase {
>
> [image: :+1:]
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58315746>
>

briancroom commented on Sources/XCTest/XCTestCase.swift:
```diff
@@ -39,8 +44,75 @@ public class XCTestCase: XCTest {
+        return XCTestCaseRun.self
+    }
+
+    public override func performTest(run: XCTestRun) {
```
Swift 3 API import change: `public func performTest(run: XCTestRun)` becomes `public func perform(run: XCTestRun)`

gruberjuliannama replied:
_name = "\(self.dynamicType).\(name)"
+ self.testClosure = testClosure
+ }
+
+ /// Invoking a test performs its setUp, invocation, and tearDown. In
+ /// general this should not be called directly.
+ public func invokeTest() {
+ setUp()
+ do {
+ try testClosure(self)
+ } catch {
+ recordFailureWithDescription(
+ "threw error \"\(error)\"",
+ inFile: "<EXPR>",
+ atLine: 0,
+ expected: false)
+ }
+ tearDown()
+ }
+
+ /// Records a failure in the execution of the test and is used by all test
+ /// assertions.
+ /// - Parameter description: The description of the failure being
reported.
+ /// - Parameter filePath: The file path to the source file where the
failure
+ /// being reported was encountered.
+ /// - Parameter lineNumber: The line number in the source file at filePath
+ /// where the failure being reported was encountered.
+ /// - Parameter expected: `true` if the failure being reported was the
+ /// result of a failed assertion, `false` if it was the result of an
+ /// uncaught exception.
+ public func recordFailureWithDescription(description: String, inFile
filePath: String, atLine lineNumber: UInt, expected: Bool) {
+ testRun?.recordFailureWithDescription(
+ "\(name) : \(description)",
+ inFile: filePath,
+ atLine: lineNumber,
+ expected: expected)
+
+ // FIXME: Apple XCTest does not throw a fatal error and crash the test
+ // process, it merely prevents the remainder of a testClosure
+ // from execting after it's been determined that it has already
+ // failed. The following behavior is incorrect.
+ // FIXME: No regression tests exist for this feature. We may break it
+ // without ever realizing.
+ if !continueAfterFailure {
+ fatalError("Terminating execution due to test failure")

2016-04-03 21:14 GMT+02:00 Brian Croom <notifications@github.com>:

> In Sources/XCTest/XCTestCase.swift
> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58315753>
> :
>
> > @@ -39,8 +44,75 @@ public class XCTestCase: XCTest {
> >      ///   https://bugs.swift.org/browse/SR-1129 for details.
> >      public var _name: String
> >
> > -    public required override init() {
> > -        _name = "\(self.dynamicType).<unknown>"
> > +    public override var testRunClass: AnyClass {
> > +        return XCTestCaseRun.self
> > +    }
> > +
> > +    public override func performTest(run: XCTestRun) {
>
> Swift 3 API import change: public func performTest(run: XCTestRun)
> becomes public func perform(run: XCTestRun)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58315753>
>

modocache replied:
Nice, thanks!

briancroom commented on Sources/XCTest/XCTestCase.swift:
```diff
@@ -39,8 +44,75 @@ public class XCTestCase: XCTest {
+    /// - Parameter expected: `true` if the failure being reported was the
+    ///   result of a failed assertion, `false` if it was the result of an
+    ///   uncaught exception.
+    public func recordFailureWithDescription(description: String, inFile filePath: String, atLine lineNumber: UInt, expected: Bool) {
```
Swift 3 API import change: 
```swift
public func recordFailureWithDescription(description: String, inFile filePath: String, atLine lineNumber: UInt, expected: Bool)
```
becomes
```swift
public func recordFailure(withDescription description: String, inFile filePath: String, atLine lineNumber: UInt, expected: Bool)
```

briancroom commented on Sources/XCTest/XCTestCase.swift:
```diff
@@ -39,8 +44,75 @@ public class XCTestCase: XCTest {
+    ///   uncaught exception.
+    public func recordFailureWithDescription(description: String, inFile filePath: String, atLine lineNumber: UInt, expected: Bool) {
+        testRun?.recordFailureWithDescription(
+            "\(name) : \(description)",
```
I'm curious about the motivation for this change (and the corresponding change in the Observation/All functional test that inspects the value of the `description` parameter eventually passed to the `XCTestObservation` failure method). Apple XCTest does not include the test name in the failure `description` parameter. In all contexts where this value is passed around, the `XCTestCase` instance itself is also available, providing access to the `name`.

modocache replied:
Nice catch. I've fixed this: observers are now passed only the description. The `PrintObserver` corelibs-xctest uses now prepends the test name itself. Thanks!

briancroom commented on Sources/XCTest/XCTestCase.swift:
```diff
@@ -82,90 +154,15 @@ extension XCTestCase {
         }
     }
 
-    internal static func invokeTests(tests: [(String, XCTestCase throws -> Void)]) {
```
I'm really happy to leave this monstrosity of a function behind!

modocache replied:
It served us well! 💐

briancroom commented on Sources/XCTest/XCTestMain.swift:
```diff
@@ -88,64 +63,30 @@ internal struct XCTRun {
-        rootTestSuites.append(XCTestSuite(name: "All tests"))
-        rootTestSuites.append(XCTestSuite(name: "\(testBundle.bundlePath.lastPathComponent).xctest"))
+        rootTestSuite = XCTestSuite(name: "All tests")
+        currentTestSuite = XCTestSuite(name: "\(testBundle.bundlePath.lastPathComponent).xctest")
```
I wonder if it's a little misleading to append `.xctest` to the output here even though we're not actually executing a test bundle? I'm not sure how I feel about this just yet. 🤔

modocache replied:
Good point. On the other hand, who knows which of the tools out there parsing XCTest output will break if they don't see a `.xctest` somewhere! 😝 

I'm not sure here either. Any opinions, @mike-ferris-apple?

briancroom commented on Sources/XCTest/XCTestMain.swift:
```diff
@@ -88,64 +63,30 @@ internal struct XCTRun {
-            observationCenter.testSuiteDidFinish(testSuite)
+    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
+        let dummyTestCase = testCaseType.init(name: "None") { _ in }
+        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")
```
I think we could skip creating the dummy test case and just stringify the `testCaseType` value itself here?

gruberjuliannama replied:
the test and is used by all test
> +    /// assertions.
> +    /// - Parameter description: The description of the failure being reported.
> +    /// - Parameter filePath: The file path to the source file where the failure
> +    ///   being reported was encountered.
> +    /// - Parameter lineNumber: The line number in the source file at filePath
> +    ///   where the failure being reported was encountered.
> +    /// - Parameter expected: `true` if the failure being reported was the
> +    ///   result of a failed assertion, `false` if it was the


2016-04-03 21:28 GMT+02:00 Brian Croom <notifications@github.com>:

> In Sources/XCTest/XCTestMain.swift
> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58316019>
> :
>
> > -        // FIXME: This was the simplest implementation that didn't involve
> > -        //        changing how swift-corelibs-xctest builds up and executes a
> > -        //        collection of test cases. However, instead of using an index
> > -        //        to enumerate both the `XCTestSuite` and the test cases at
> > -        //        once, we should enumerate the test cases in the `XCTestSuite`.
> > -        //        `XCTestSuite` should be responsible for representing the
> > -        //        collection of tests we execute here.
> > -        for (index, test) in filteredTestCases.enumerated() {
> > -            let (testCase, tests) = test
> > -            let testSuite = rootTestSuites.last!.tests[index] as! XCTestSuite
> > -            observationCenter.testSuiteWillStart(testSuite)
> > -            testCase.invokeTests(tests)
> > -            observationCenter.testSuiteDidFinish(testSuite)
> > +    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
> > +        let dummyTestCase = testCaseType.init(name: "None") { _ in }
> > +        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")
>
> I think we could skip creating the dummy test case and just stringify the
> testCaseType value itself here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58316019>
>

gruberjuliannama replied:
simplest implementation that didn't involve
> -        //        changing how swift-corelibs-xctest builds up and executes a
> -        //        collection of test cases. However, instead of using an index
> -        //        to enumerate both the `XCTestSuite` and the test cases at
> -        //        once, we should enumerate the test cases in the `XCTestSuite`.
> -        //        `XCTestSuite` should be responsible for representing the
> -        //        collection of tests we execute here.
> -        for (index, test) in filteredTestCases.enumerated() {
> -            let (testCase, tests) = test
> -            let testSuite = rootTestSuites.last!.tests[index] as! XCTestSuite
> -            observationCenter.testSuiteWillStart(testSuite)
> -            testCase.invokeTests(tests)
> -            observationCenter.testSuiteDidFinish(testSuite)
> +    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
> +        let dummyTestCase = testCaseType.init(name: "None") { _ in }
> +        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")


2016-04-03 21:29 GMT+02:00 Julian Gruber <julaner007@gmail.com>:

> the test and is used by all test
> > +    /// assertions.
> > +    /// - Parameter description: The description of the failure being reported.
> > +    /// - Parameter filePath: The file path to the source file where the failure
> > +    ///   being reported was encountered.
> > +    /// - Parameter lineNumber: The line number in the source file at filePath
> > +    ///   where the failure being reported was encountered.
> > +    /// - Parameter expected: `true` if the failure being reported was the
> > +    ///   result of a failed assertion, `false` if it was the
>
>
> 2016-04-03 21:28 GMT+02:00 Brian Croom <notifications@github.com>:
>
>> In Sources/XCTest/XCTestMain.swift
>> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58316019>
>> :
>>
>> > -        // FIXME: This was the simplest implementation that didn't involve
>> > -        //        changing how swift-corelibs-xctest builds up and executes a
>> > -        //        collection of test cases. However, instead of using an index
>> > -        //        to enumerate both the `XCTestSuite` and the test cases at
>> > -        //        once, we should enumerate the test cases in the `XCTestSuite`.
>> > -        //        `XCTestSuite` should be responsible for representing the
>> > -        //        collection of tests we execute here.
>> > -        for (index, test) in filteredTestCases.enumerated() {
>> > -            let (testCase, tests) = test
>> > -            let testSuite = rootTestSuites.last!.tests[index] as! XCTestSuite
>> > -            observationCenter.testSuiteWillStart(testSuite)
>> > -            testCase.invokeTests(tests)
>> > -            observationCenter.testSuiteDidFinish(testSuite)
>> > +    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
>> > +        let dummyTestCase = testCaseType.init(name: "None") { _ in }
>> > +        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")
>>
>> I think we could skip creating the dummy test case and just stringify the
>> testCaseType value itself here?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58316019>
>>
>
>

modocache replied:
Wow, that worked! Thanks!

briancroom commented on Sources/XCTest/XCTestRun.swift:
```diff
@@ -0,0 +1,150 @@
+    /// The number of seconds that elapsed between when the run was started and
+    /// when it was stopped.
+    public var totalDuration: NSTimeInterval {
+        guard stopped else {
```
How would you feel about doing optional unwrapping here to avoid the need for the `!`s for the return statement?

modocache replied:
Good call, I forgot we can unwrap multiple optionals at once now. I'll change this.

briancroom commented on Sources/XCTest/XCTestRun.swift:
```diff
@@ -0,0 +1,150 @@
+    /// - Parameter expected: `true` if the failure being reported was the
+    ///   result of a failed assertion, `false` if it was the result of an
+    ///   uncaught exception.
+    public func recordFailureWithDescription(description: String, inFile filePath: String, atLine lineNumber: UInt, expected: Bool) {
```
Swift 3 API import change: 
```swift
public func recordFailureWithDescription(description: String, inFile filePath: String?, atLine lineNumber: UInt, expected: Bool)
``` 
becomes 
```swift
public func recordFailure(withDescription description: String, inFile filePath: String?, atLine lineNumber: UInt, expected: Bool)
```

briancroom commented on Sources/XCTest/XCTestSuiteRun.swift:
```diff
@@ -0,0 +1,67 @@
+
+    /// Add a test run to the collection of `testRuns`.
+    /// - Note: It is rare to call this method outside of XCTest itself.
+    public func addTestRun(testRun: XCTestRun) {
```
Swift 3 API import change: `public func addTestRun(testRun: XCTestRun)` becomes `public func addTest(testRun: XCTestRun)`

briancroom commented on Tests/Functional/Observation/All/main.swift:
```diff
@@ -74,10 +78,10 @@ class Observation: XCTestCase {
 
         XCTFail("fail!")
-        XCTAssertEqual(observer.failureDescriptions, ["failed - fail!"])
+        XCTAssertEqual(observer.failureDescriptions, ["Observation.test_one : failed - fail!"])
```
This is the test change I was referring to in the comment [above](https://github.com/apple/swift-corelibs-xctest/pull/86/files#r58315910).

gruberjuliannama replied:
break
}
// Otherwise, wait another fraction of a second.
runLoop.runUntilDate(NSDate(timeIntervalSinceNow: 0.01))
} while NSDate().compare(timeoutDate) == NSComparisonResult.OrderedAscending
if unfulfilledDescriptions.count > 0 {
// Not all expectations were fulfilled. Append a failure
// to the array of expectation-based failures.
let descriptions = unfulfilledDescriptions.joined(separator: ", ")
- let failure = XCTFailure(
- message: "Exceeded timeout of \(timeout) seconds, with unfulfilled
expectations: \(descriptions)",
- failureDescription: "Asynchronous wait failed",
- expected: true,
- file: file,
- line: line)
- if let failureHandler = XCTFailureHandler {
- failureHandler(failure)

2016-04-03 21:32 GMT+02:00 Brian Croom <notifications@github.com>:

> In Tests/Functional/Observation/All/main.swift
> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58316115>
> :
>
> > @@ -74,10 +78,10 @@ class Observation: XCTestCase {
> >          XCTAssertEqual(observer.finishedBundlePaths.count, 0)
> >
> >          XCTFail("fail!")
> > -        XCTAssertEqual(observer.failureDescriptions, ["failed - fail!"])
> > +        XCTAssertEqual(observer.failureDescriptions, ["Observation.test_one : failed - fail!"])
>
> This is the test change I was referring to in the comment above
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files#r58315910>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58316115>
>

briancroom commented:
Wow, this is really great @modocache! My biggest complaint is that you've left so little API parity work for others (i.e. me) to have fun with 😝. In all seriousness though, I'm happy to see this, and it would have been very difficult to break it into more PRs without the smaller steps being pretty meaningless on their own.

I've left a handful of minor comments and questions for you!

gruberjuliannama commented:
twitter, knn = Twitter(), KNN()
>>>
>>> for i in range(1, 10):
>>>     for tweet in twitter.search('#win OR #fail', start=i, count=100):
>>>         s = tweet.text.lower()
>>>         p = '#win' in s and 'WIN' or 'FAIL'
>>>         v = tag(s)
>>>         v = [word for word, pos in v if pos == 'JJ'] # JJ = adjective
>>>         v = count(v)
>>>         if v:
>>>             knn.train(v, type=p)
>>>
>>> print knn.classify('sweet potato burger')
>>> print knn.classify('stupid autocorrect')

'WIN'
'FAIL'

2016-04-03 21:35 GMT+02:00 Brian Croom <notifications@github.com>:

> Wow, this is really great @modocache <https://github.com/modocache>! My
> biggest complaint is that you've left so little API parity work for others
> (i.e. me) to have fun with 😝. In all seriousness though, I'm happy to
> see this, and it would have been very difficult to break it into more PRs
> without the smaller steps being pretty meaningless on their own.
>
> I've left a handful of minor comments and questions for you!
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86#issuecomment-205040643>
>

briancroom commented on Sources/XCTest/XCTestCase.swift:
```diff
@@ -39,8 +44,75 @@ public class XCTestCase: XCTest {
 
-    public required override init() {
-        _name = "\(self.dynamicType).<unknown>"
+    public override var testRunClass: AnyClass {
```
`XCTestCase` is missing an override for `testCaseCount` (AFAICT it should just always `return 1`)

modocache replied:
Wow, nice catch!! Thanks.

briancroom commented on Sources/XCTest/XCTestRun.swift:
```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    private var started: Bool {
```
I think these would become `isStarted` and `isStopped` to follow the Swift 3 API guidelines.

modocache replied:
Done! Also, I think I found a Swift bug. After changed these to `isStarted` and `isStopped`, I tried rebuilding the project and it failed:

```
d /Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Products/Debug/SwiftXCTest.framework/Versions/A/SwiftXCTest normal x86_64
    cd /Users/bgesiak/GitHub/apple/swift-corelibs-xctest
    export MACOSX_DEPLOYMENT_TARGET=10.11
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -arch x86_64 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -L/Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Products/Debug -F/Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Products/Debug -filelist /Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Intermediates/XCTest.build/Debug/SwiftXCTest.build/Objects-normal/x86_64/SwiftXCTest.LinkFileList -install_name @rpath/SwiftXCTest.framework/Versions/A/SwiftXCTest -Xlinker -rpath -Xlinker @executable_path/../Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -mmacosx-version-min=10.11 -Xlinker -no_deduplicate -fobjc-link-runtime -L/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-03-24-a.xctoolchain/usr/lib/swift/macosx -Xlinker -add_ast_path -Xlinker /Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Intermediates/XCTest.build/Debug/SwiftXCTest.build/Objects-normal/x86_64/SwiftXCTest.swiftmodule -framework SwiftFoundation -single_module -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Intermediates/XCTest.build/Debug/SwiftXCTest.build/Objects-normal/x86_64/SwiftXCTest_dependency_info.dat -o /Users/bgesiak/Library/Developer/Xcode/DerivedData/XCTest-gdiiwlsynrssmkdgamrgfoqbjewl/Build/Products/Debug/SwiftXCTest.framework/Versions/A/SwiftXCTest

Undefined symbols for architecture x86_64:
  "SwiftXCTest.XCTestRun.(started in _591E1E8C99D08309B8F88D133866F0A8).getter : Swift.Bool", referenced from:
      type metadata for SwiftXCTest.XCTestCaseRun in XCTestCaseRun.o
      type metadata for SwiftXCTest.XCTestSuiteRun in XCTestSuiteRun.o
  "SwiftXCTest.XCTestRun.(stopped in _591E1E8C99D08309B8F88D133866F0A8).getter : Swift.Bool", referenced from:
      type metadata for SwiftXCTest.XCTestCaseRun in XCTestCaseRun.o
      type metadata for SwiftXCTest.XCTestSuiteRun in XCTestSuiteRun.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Clearing out derived data solved the problem. Looks like incremental builds are kind of broken when only modifying private members?

modocache commented:
Haha, thanks @briancroom. Although, even aside from `NSInvocation`-related methods, performance testing, and XCUITest, there are still some API discrepancies. I was looking through [the Swift headers for Apple XCTest](https://github.com/modocache/XCTest.swift/blob/5b5a152150bacea9c373dcfb16a8384563cd4419/XCTest.swift) and found several public API we don't provide or have discrepancies with:

```swift

var XCT_GENERICS_AVAILABLE: Int32 { get }
var XCT_NULLABLE_AVAILABLE: Int32 { get }
var XCT_UI_TESTING_AVAILABLE: Int32 { get }

let XCTestErrorDomain: String
enum XCTestErrorCode : Int {
  init?(rawValue rawValue: Int)
  var rawValue: Int { get }
  case timeoutWhileWaiting
  case failureWhileWaiting
}

class XCTestCase : XCTest {
  init(selector selector: Selector)
}

extension XCTestCase {
  // We define these two methods, but no longer match the Swift 3 signature.
  func expectation(withDescription description: String) -> XCTestExpectation
  func waitForExpectations(withTimeout timeout: NSTimeInterval, handler handler: XCWaitCompletionHandler? = nil)

  func keyValueObservingExpectation(for objectToObserve: AnyObject, keyPath keyPath: String, expectedValue expectedValue: AnyObject?) -> XCTestExpectation
  func keyValueObservingExpectation(for objectToObserve: AnyObject, keyPath keyPath: String, handler handler: XCKeyValueObservingExpectationHandler? = nil) -> XCTestExpectation
  func expectation(for predicate: NSPredicate, evaluatedWith object: AnyObject, handler handler: XCPredicateExpectationHandler? = nil) -> XCTestExpectation
}
typealias XCKeyValueObservingExpectationHandler = (AnyObject, [NSObject : AnyObject]) -> Bool
typealias XCPredicateExpectationHandler = () -> Bool

class XCTestSuite : XCTest {
  class func defaultTestSuite() -> Self
  convenience init(forBundlePath bundlePath: String)
  convenience init(forTestCaseWithName name: String)
  convenience init(forTestCaseClass testCaseClass: AnyClass)
}
```

I'll create an issue to track these once this is merged. Speaking of which, I think I've addressed all feedback so far--what do you think, @mike-ferris-apple?

modocache commented:
I rebased this onto @briancroom's changes from #88. @mike-ferris-apple, care to kick off CI for this? I'd like to merge today in order to unblock @harlanhaskins in #91.

harlanhaskins commented:
Thanks, @modocache!

mike-ferris-apple commented:
@swift-ci please test

modocache commented:
Looks like a linker failure when building for SwiftPM. Will take a look tonight.

harlanhaskins commented:
@modocache If it's alright with you, I'd like to fix and merge #91 today.

modocache commented:
@harlanhaskins Sure thing!

modocache commented:
Fixed the build failure, but let's hold off on merging this until after @harlanhaskins is done with #92.

harlanhaskins commented:
Well, we're now waiting for conflicts to resolve, so if you want to merge this wholesale now, I can update my PR.

harlanhaskins commented:
Also, @modocache looking at the tests, there seem to be a lot of places where the line numbers are the only thing that change. It might be worthwhile to update them to use `[[@LINE+<offset>]]` instead.

modocache commented:
@harlanhaskins We aren't using `FileCheck`, and our checker doesn't support offsets yet. Support is added in #68 but haven't landed due to some incredible merge conflicts. Will fix soon!

@mike-ferris-apple One last Swift CI test to merge?

harlanhaskins commented:
Oops, didn't notice it wasn't FileCheck; my apologies!

mike-ferris-apple commented:
@swift-ci please test

modocache commented:
The test passed before. Just rebased onto #91. @mike-ferris-apple care to run CI again?

mike-ferris-apple commented:
@swift-ci please test

modocache commented:
Woohoo! 🎉 

So, so close to 100% API parity (modulo performance tests, UI tests, and `NSInvocation`-related stuff)!

briancroom commented:
Excellent! Thanks for your effort on this @modocache! :100:

modocache commented:
> My biggest complaint is that you've left so little API parity work for others (i.e. me) to have fun with 😝

@briancroom [SR-1355](https://bugs.swift.org/browse/SR-1355) is calling out to you... 😉

briancroom commented:
@modocache I've already started digging in to see all the failures it can throw when you misuse the API 😆
